html/captive-portal/lib/captiveportal/PacketFence/Controller/TLSProfile.pm

```diff
@@ -0,0 +1,242 @@
+    my $mac = $c->portalSession->clientMac;
+    my $node_info = node_view($mac);
+    my $pid = $node_info->{'pid'};
+    my $provisioner = $c->profile->findProvisioner($mac);
+    #$provisioner->authorize($mac) if (defined($provisioner));
+    my $certificate_cn = $mac;
+    $certificate_cn =~ s/:/-/g;
+    $c->stash(
+        post_uri            => '/tlsprofile/cert_process',
+        certificate_cn      => $certificate_cn,
+        certificate_pwd     => $request->param_encoded("certificate_pwd"),
+        certificate_pwd_check     => $request->param_encoded("certificate_pwd_check"),
+        certificate_email   => lc( $request->param_encoded("certificate_email") || $request->param_encoded("email")),
+        template            => 'pki.html',
+        provisioner         => $provisioner,
+        username            => $username,
+        mac                 => $mac,
+        pid                 => $pid,
+    );
+}
+
+sub build_cert_p12 : Path : Args(0) {
+    my ($self, $c) = @_;
+    my $logger = $c->log;
+    my $session = $c->session;
+    my $cert_data = $c->stash->{'cert_content'};
+    my $cert = $c->stash->{'certificate_cn'} . ".p12";
+    my $certname = "$cert_dir/$cert";
+    $c->session( certificate_cn => "$cert" );
+    my $portalSession = $c->portalSession;
+    my $mac           = $portalSession->clientMac;
+    my $node_info     = node_view($mac);
+    my $pid           = $node_info->{'pid'};
+    my $fh;
+    open ($fh, '>', $certname);
+    if (-e $certname) {
+        $logger->info("Certificate for user \"$pid\" successfully created.");
+    }
+    else {
+        $logger->info("The certificate file could not be saved for username \"$pid\"");
+        $self->showError($c,"An error has occured while trying to save your certificate, please contact your IT support");
+    }
+    print $fh "$cert_data\n";
+    close $fh;
+}
+
+sub get_cert : Private {
+    my ($self, $c) = @_;
+    my $portalSession = $c->portalSession;
+    my $stash = $c->stash;
+    my $mac           = $portalSession->clientMac;
+    my $provisioner   = $c->profile->findProvisioner($mac);
+    return unless $provisioner;
+    my $pki_provider = $provisioner->getPkiProvider();
+    my $cert_content = $pki_provider->get_cert({ certificate_email => $stash->{certificate_email}, certificate_cn => $stash->{certificate_cn}, certificate_pwd => $stash->{certificate_pwd} });
+    $c->log->debug("cert_content from pki service $cert_content");
+    $c->stash(cert_content => $cert_content);
+}
+
+sub cert_process : Local {
+    my ($self,$c) = @_;
+    my $logger = $c->log;
+    $c->stash(info => $c->session->{info});
+    $c->forward('validateform');
+    $c->forward('get_cert');
+    $c->forward('build_cert_p12');
+    $c->forward('b64_cert');
+    $c->forward('export_fingerprint');
+    $c->forward( 'Authenticate' => 'checkIfProvisionIsNeeded' );
+    $c->forward( 'CaptivePortal' => 'webNodeRegister', [$c->stash->{info}{pid}, %{$c->stash->{info}}]);
```

Comment:
Make sure Authenticate => postAuthentication is called in your flow somewhere because it handles important things.